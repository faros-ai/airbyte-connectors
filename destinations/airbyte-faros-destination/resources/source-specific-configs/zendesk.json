{
  "title": "Zendesk",
  "type": "object",
  "oneOf": [
    {
      "title": "Configuration",
      "type": "object",
      "properties": {
        "source_type": {
          "type": "string",
          "const": "Zendesk",
          "order": 0
        },
        "ticket_additional_fields": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "title": "Ticket Additional Fields",
          "description": "The additional field names from the ticket to include in the task record.",
          "order": 1
        },
        "additional_fields_array_limit": {
          "type": "integer",
          "title": "Additional Fields Array Limit",
          "description": "Truncates an additional field's array value to the given length.",
          "default": 50,
          "order": 2
        },
        "team_mapping": {
          "type": "string",
          "title": "Team Mapping",
          "description": "JSON map of Zendesk group(s) name, to org_Team uids.",
          "multiline": true,
          "default": "{}",
          "examples": [
            "{ \"Group 1\": \"team1\", \"*\": \"all_teams\" }"
          ],
          "order": 3
        },
        "sync_groups": {
          "type": "boolean",
          "title": "Sync Groups",
          "description": "Store groups as org_Teams based on group name",
          "default": false,
          "order": 4
        }
      }
    }
  ]
}
