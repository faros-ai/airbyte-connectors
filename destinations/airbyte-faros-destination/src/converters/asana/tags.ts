import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AsanaConverter} from './common';

export class Tags extends AsanaConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_Label'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const tag = record.record.data;

    return [
      {
        model: 'tms_Label',
        record: {
          name: tag.name,
        },
      },
    ];
  }
}
