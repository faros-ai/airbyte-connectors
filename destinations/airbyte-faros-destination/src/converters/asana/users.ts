import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AsanaCommon, AsanaConverter, AsanaUser} from './common';

export class Users extends AsanaConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as AsanaUser;

    return [AsanaCommon.tms_User(user, source)];
  }
}
