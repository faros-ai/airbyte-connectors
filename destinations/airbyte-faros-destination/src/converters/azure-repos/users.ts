import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureReposConverter} from './common';
import {User, UserType, UserTypeCategory} from './models';

export class Users extends AzureReposConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'vcs_Membership',
    'vcs_User',
  ];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];
    const organizationName = this.getOrganizationFromUrl(
      userItem._links.self.href
    );
    const organization = {uid: organizationName, source};
    const type: UserType = {
      category: UserTypeCategory.User,
      detail: userItem.subjectKind,
    };
    res.push({
      model: 'vcs_Membership',
      record: {
        organization,
        user: {uid: userItem.principalName, source},
      },
    });

    res.push({
      model: 'vcs_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        type,
        htmlUrl: userItem.url,
        source,
      },
    });
    return res;
  }
}
