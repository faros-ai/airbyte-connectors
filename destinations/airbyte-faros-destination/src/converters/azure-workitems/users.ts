import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureWorkitemsConverter} from './common';
import {User} from './models';

export class Users extends AzureWorkitemsConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];

    res.push({
      model: 'tms_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        emailAddress: userItem.mailAddress,
        source,
      },
    });
    return res;
  }
}
