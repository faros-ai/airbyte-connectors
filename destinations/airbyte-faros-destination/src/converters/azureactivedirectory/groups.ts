import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureActiveDirectoryConverter} from './common';
import {Group} from './models';

export class Groups extends AzureActiveDirectoryConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'org_Team',
    'org_TeamMembership',
  ];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const group = record.record.data as Group;
    const res: DestinationRecord[] = [];
    const uid = group.id;
    const lead =
      group.owners.length >= 1
        ? {
            uid: group.owners[0],
            source,
          }
        : undefined;

    res.push({
      model: 'org_Team',
      record: {
        uid,
        name: group.displayName,
        description: group.description,
        lead,
        parentTeam: {uid: 'all_teams'}, // TODO: compute parent team
        teamChain: ['all_teams'], // TODO: team chain
        source,
      },
    });

    for (const user of group.members ?? []) {
      res.push({
        model: 'org_TeamMembership',
        record: {
          team: {uid},
          member: {uid: user},
        },
      });
    }
    return res;
  }
}
