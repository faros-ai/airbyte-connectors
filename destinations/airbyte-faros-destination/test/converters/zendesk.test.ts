import {AirbyteLogger, AirbyteRecord} from 'faros-airbyte-cdk';
import _ from 'lodash';
import {getLocal} from 'mockttp';

import {StreamContext} from '../../src/converters/converter';
import {
  GroupsStream,
  TicketFieldsStream,
  TicketMetricsStream,
} from '../../src/converters/zendesk/common';
import {Groups} from '../../src/converters/zendesk/groups';
import {SatisfactionRatings} from '../../src/converters/zendesk/satisfaction_ratings';
import {Tags} from '../../src/converters/zendesk/tags';
import {Tickets} from '../../src/converters/zendesk/tickets';
import {Users} from '../../src/converters/zendesk/users';
import {initMockttp, tempConfig} from '../testing-tools';
import {destinationWriteTest} from './utils';

describe('zendesk', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let configPath: string;

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig({
      api_url: mockttp.url,
      source_specific_configs: {zendesk: {sync_groups: true}},
    });
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    await destinationWriteTest({
      configPath,
      catalogPath: 'test/resources/zendesk/catalog.json',
      inputRecordsPath: 'zendesk/all-streams.log',
    });
  });
});

describe('tickets', () => {
  const converter = new Tickets();
  const ctx = new StreamContext(
    new AirbyteLogger(),
    {
      edition_configs: {},
      source_specific_configs: {
        zendesk: {
          ticket_additional_fields: [
            'Affected Systems',
            'Needs More Triage',
            'Agent internal comments',
            'Reply customer on',
          ],
        },
      },
    },
    {}
  );

  ctx.set(
    TicketFieldsStream.asString,
    '1',
    AirbyteRecord.make('ticket_fields', {
      id: 1,
      type: 'multiselect',
      title: 'Affected Systems',
    })
  );
  ctx.set(
    TicketFieldsStream.asString,
    '2',
    AirbyteRecord.make('ticket_fields', {
      id: 2,
      type: 'checkbox',
      title: 'Needs More Triage',
    })
  );
  ctx.set(
    TicketFieldsStream.asString,
    '3',
    AirbyteRecord.make('ticket_fields', {
      id: 3,
      type: 'text',
      title: 'Agent internal comments',
    })
  );
  ctx.set(
    TicketFieldsStream.asString,
    '4',
    AirbyteRecord.make('ticket_fields', {
      id: 4,
      type: 'date',
      title: 'Reply customer on',
    })
  );
  ctx.set(
    TicketFieldsStream.asString,
    '30',
    AirbyteRecord.make('ticket_fields', {
      id: 30,
      type: 'integer',
      title: 'Version impacted',
    })
  );
  ctx.set(
    TicketFieldsStream.asString,
    '6',
    AirbyteRecord.make('ticket_fields', {
      title: 'Ticket status',
      id: 6,
      type: 'custom_status',
      custom_statuses: [
        {
          id: 23328843804692,
          status_category: 'pending',
          agent_label: 'Waiting for fix',
          end_user_label: 'Waiting for fix',
          description: 'There is nothing to do for this ticket',
        },
      ],
    })
  );
  const ticket = {
    url: 'https://example.zendesk.com/api/v2/tickets/1.json',
    id: 1,
    external_id: null,
    via: {
      channel: 'sample_ticket',
      source: {from: {}, to: {}, rel: null},
    },
    created_at: '2024-01-24T14:15:15Z',
    updated_at: '2024-01-24T14:15:15Z',
    type: 'incident',
    subject: 'Sample ticket: Meet the ticket',
    raw_subject: 'Sample ticket: Meet the ticket',
    description:
      'Hi there, I’m sending an email because I’m having a problem setting up your new product',
    priority: 'normal',
    status: 'open',
    recipient: null,
    requester_id: 23111599474068,
    submitter_id: 23111595978772,
    assignee_id: null,
    organization_id: null,
    group_id: 23111598987540,
    collaborator_ids: [],
    follower_ids: [],
    email_cc_ids: [],
    forum_topic_id: null,
    problem_id: null,
    has_incidents: false,
    is_public: true,
    due_at: '2024-02-01T00:00:00Z',
    tags: [],
    custom_fields: [],
    satisfaction_rating: null,
    sharing_agreement_ids: [],
    custom_status_id: 23111598975892,
    fields: [{id: 23111693329300, value: null}],
    followup_ids: [],
    ticket_form_id: 23111598732820,
    brand_id: 23111567404180,
    allow_channelback: false,
    allow_attachments: true,
    from_messaging_channel: false,
    generated_timestamp: 1706105716,
  };

  test('ticket', async () => {
    const record = AirbyteRecord.make('ticket', ticket);
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('tags', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      tags: ['tag1', 'tag2'],
    });
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('assignee', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      assignee_id: 23123038291476,
    });
    const assigneeCtx = new StreamContext(
      new AirbyteLogger(),
      {edition_configs: {}},
      {}
    );
    assigneeCtx.set(
      TicketMetricsStream.asString,
      'metricId',
      AirbyteRecord.make('ticket_metrics', {
        ticket_id: ticket.id,
        assigned_at: '2024-01-24T14:17:58Z',
      })
    );
    const res = await converter.convert(record, assigneeCtx);
    expect(res).toMatchSnapshot();
  });

  test('solved', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      status: 'solved',
    });
    const metricCtx = new StreamContext(
      new AirbyteLogger(),
      {edition_configs: {}},
      {}
    );
    metricCtx.set(
      TicketMetricsStream.asString,
      'metricId',
      AirbyteRecord.make('ticket_metrics', {
        ticket_id: ticket.id,
        solved_at: '2024-01-24T20:12:02Z',
      })
    );
    const res = await converter.convert(record, metricCtx);
    expect(res).toMatchSnapshot();
  });

  test('follow ups', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      followup_ids: [2],
    });
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('custom status', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      status: 'pending',
      custom_status_id: 23328843804692,
    });
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('additional fields', async () => {
    const record = AirbyteRecord.make('ticket', {
      ...ticket,
      custom_fields: [
        {id: 1, value: ['api', 'ui']},
        {id: 2, value: true},
        {id: 3, value: 'No extra work needed'},
        {id: 4, value: null},
        {id: 30, value: 3},
      ],
    });
    const res = await converter.convert(record, ctx);
    const taskRec = res.find((r) => r.model === 'tms_Task');
    expect(taskRec.record.additionalFields).toHaveLength(4);
    expect(res).toMatchSnapshot();
  });
});

describe('tags', () => {
  const converter = new Tags();
  const tag = {name: 'support', count: 1};
  test('tag', async () => {
    const record = AirbyteRecord.make('tag', tag);

    const res = await converter.convert(record);
    expect(res).toMatchSnapshot();
  });
});

describe('users', () => {
  const converter = new Users();
  const user = {
    id: 111,
    url: 'https://example.zendesk.com/api/v2/users/111.json',
    name: 'The Customer',
    email: 'customer@example.com',
    created_at: '2024-01-24T14:17:56Z',
    updated_at: '2024-01-24T14:17:56Z',
    active: true,
  };

  test('user', async () => {
    const record = AirbyteRecord.make('user', user);
    const res = await converter.convert(record);
    expect(res).toMatchSnapshot();
  });

  test('suspended user', async () => {
    const record = AirbyteRecord.make('user', {
      ...user,
      suspended: true,
    });
    const res = await converter.convert(record);
    expect(res).toMatchSnapshot();
  });
});

describe('satisfaction ratings', () => {
  const converter = new SatisfactionRatings();

  const ctx = new StreamContext(new AirbyteLogger(), {edition_configs: {}}, {});
  ctx.set(
    GroupsStream.asString,
    '101',
    AirbyteRecord.make('groups', {
      id: 101,
      name: 'Group 101',
      description: 'Group 101',
    })
  );

  const rating = {
    id: 1,
    ticket_id: 15,
    score: 'offered',
    created_at: '2024-02-07T21:51:23Z',
    updated_at: '2024-02-07T21:51:23Z',
    comment: null,
    group_id: 101,
  };

  test('rating', async () => {
    const record = AirbyteRecord.make('satisfaction_rating', rating);

    const res = await converter.convert(record, ctx);
    expect(res).toHaveLength(3);
    expect(res).toMatchSnapshot();
  });

  test('fallback to group when no group in context', async () => {
    const fallbackCtx = new StreamContext(
      new AirbyteLogger(),
      {
        edition_configs: {},
        source_specific_configs: {
          zendesk: {
            team_mapping: {'*': 'all_teams'},
          },
        },
      },
      {}
    );
    const record = AirbyteRecord.make('satisfaction_rating', rating);

    const res = await converter.convert(record, fallbackCtx);
    expect(res).toHaveLength(2);
    expect(res).toMatchSnapshot();
  });
});

describe('groups', () => {
  const ctx = new StreamContext(
    new AirbyteLogger(),
    {
      edition_configs: {},
      source_specific_configs: {
        zendesk: {
          sync_groups: true,
          team_mapping: {
            '*': 'all_teams',
            'Group 1': 'Team 1',
          },
        },
      },
    },
    {}
  );

  const converter = new Groups();
  const group = {id: 1, name: 'Group 1', description: 'Group 1'};

  test('group', async () => {
    const teamCtx = new StreamContext(
      new AirbyteLogger(),
      {
        edition_configs: {},
        source_specific_configs: {
          zendesk: {sync_groups: true},
        },
      },
      {}
    );
    const record = AirbyteRecord.make('group', group);
    const res = await converter.convert(record, teamCtx);
    expect(res).toMatchSnapshot();
  });

  test('team mapping', async () => {
    const record = AirbyteRecord.make('group', group);
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('fallback team', async () => {
    const record = AirbyteRecord.make('group', {...group, name: 'Group 2'});
    const res = await converter.convert(record, ctx);
    expect(res).toMatchSnapshot();
  });

  test('syncing disabled', async () => {
    const syncCtx = new StreamContext(
      new AirbyteLogger(),
      {
        edition_configs: {},
        source_specific_configs: {
          zendesk: {sync_groups: false},
        },
      },
      {}
    );

    const record = AirbyteRecord.make('group', {...group, name: 'Group 3'});
    const res = await converter.convert(record, syncCtx);
    expect(res).toHaveLength(0);
  });
});
