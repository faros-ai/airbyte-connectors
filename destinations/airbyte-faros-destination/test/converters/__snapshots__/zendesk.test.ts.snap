// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groups fallback team 1`] = `
Array [
  Object {
    "model": "org_Team",
    "record": Object {
      "uid": "all_teams",
    },
  },
]
`;

exports[`groups group 1`] = `
Array [
  Object {
    "model": "org_Team",
    "record": Object {
      "description": "Group 1",
      "name": "Group 1",
      "uid": "Group 1",
    },
  },
]
`;

exports[`groups team mapping 1`] = `
Array [
  Object {
    "model": "org_Team",
    "record": Object {
      "uid": "Team 1",
    },
  },
]
`;

exports[`satisfaction ratings fallback to group when no group in context 1`] = `
Array [
  Object {
    "model": "faros_MetricValue",
    "record": Object {
      "computedAt": 2024-02-07T21:51:23.000Z,
      "definition": Object {
        "uid": "zendesk-satisfaction-ratings",
      },
      "uid": "zendesk-satisfaction-rating-1",
      "value": "offered",
    },
  },
  Object {
    "model": "org_TeamMetric",
    "record": Object {
      "team": Object {
        "uid": "all_teams",
      },
      "value": Object {
        "definition": Object {
          "uid": "zendesk-satisfaction-ratings",
        },
        "uid": "zendesk-satisfaction-rating-1",
      },
    },
  },
]
`;

exports[`satisfaction ratings rating 1`] = `
Array [
  Object {
    "model": "faros_MetricDefinition",
    "record": Object {
      "name": "Zendesk Satisfaction Ratings",
      "uid": "zendesk-satisfaction-ratings",
      "valueSource": Object {
        "category": "MetricValueEntries",
      },
      "valueType": Object {
        "category": "String",
      },
    },
  },
  Object {
    "model": "faros_MetricValue",
    "record": Object {
      "computedAt": 2024-02-07T21:51:23.000Z,
      "definition": Object {
        "uid": "zendesk-satisfaction-ratings",
      },
      "uid": "zendesk-satisfaction-rating-1",
      "value": "offered",
    },
  },
  Object {
    "model": "org_TeamMetric",
    "record": Object {
      "team": Object {
        "uid": "Group 101",
      },
      "value": Object {
        "definition": Object {
          "uid": "zendesk-satisfaction-ratings",
        },
        "uid": "zendesk-satisfaction-rating-1",
      },
    },
  },
]
`;

exports[`tags tag 1`] = `
Array [
  Object {
    "model": "tms_Label",
    "record": Object {
      "name": "support",
    },
  },
]
`;

exports[`tickets additional fields 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [
        Object {
          "name": "Affected Systems",
          "value": "api,ui",
        },
        Object {
          "name": "Needs More Triage",
          "value": "true",
        },
        Object {
          "name": "Agent internal comments",
          "value": "No extra work needed",
        },
        Object {
          "name": "Reply customer on",
          "value": null,
        },
      ],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "open",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets assignee 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "open",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskAssignment",
    "record": Object {
      "assignedAt": 2024-01-24T14:17:58.000Z,
      "assignee": Object {
        "source": "Zendesk",
        "uid": "23123038291476",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets custom status 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "Waiting for fix",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets follow ups 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "open",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskDependency",
    "record": Object {
      "dependencyType": Object {
        "category": "CreatedBy",
      },
      "dependentTask": Object {
        "source": "Zendesk",
        "uid": "2",
      },
      "fulfillingTask": Object {
        "source": "Zendesk",
        "uid": "1",
      },
      "fulfillingType": Object {
        "category": "Created",
      },
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets solved 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "resolutionStatus": "solved",
      "resolvedAt": 2024-01-24T20:12:02.000Z,
      "source": "Zendesk",
      "status": Object {
        "category": "Done",
        "detail": "solved",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets tags 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "open",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskTag",
    "record": Object {
      "label": Object {
        "name": "tag1",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskTag",
    "record": Object {
      "label": Object {
        "name": "tag2",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
]
`;

exports[`tickets ticket 1`] = `
Array [
  Object {
    "model": "tms_Task",
    "record": Object {
      "additionalFields": Array [],
      "createdAt": 2024-01-24T14:15:15.000Z,
      "creator": Object {
        "source": "Zendesk",
        "uid": "23111595978772",
      },
      "description": "Hi there, I’m sending an email because I’m having a problem setting up your new product",
      "dueAt": 2024-02-01T00:00:00.000Z,
      "name": "Sample ticket: Meet the ticket",
      "priority": "normal",
      "source": "Zendesk",
      "status": Object {
        "category": "InProgress",
        "detail": "open",
      },
      "statusChangedAt": undefined,
      "type": Object {
        "category": "SupportCase",
        "detail": "incident",
      },
      "uid": "1",
      "updatedAt": 2024-01-24T14:15:15.000Z,
      "url": "https://example.zendesk.com/api/v2/tickets/1.json",
    },
  },
  Object {
    "model": "tms_TaskBoardRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_TaskProjectRelationship",
    "record": Object {
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "task": Object {
        "source": "Zendesk",
        "uid": "1",
      },
    },
  },
  Object {
    "model": "tms_Project",
    "record": Object {
      "name": "Zendesk Tickets",
      "source": "Zendesk",
      "uid": "zendesk_tickets",
    },
  },
  Object {
    "model": "tms_TaskBoard",
    "record": Object {
      "name": "Zendesk Tickets",
      "source": "Zendesk",
      "uid": "zendesk_tickets",
    },
  },
  Object {
    "model": "tms_TaskBoardProjectRelationship",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "project": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
    },
  },
  Object {
    "model": "faros_TmsTaskBoardOptions",
    "record": Object {
      "board": Object {
        "source": "Zendesk",
        "uid": "zendesk_tickets",
      },
      "inclusion": Object {
        "category": "Included",
      },
    },
  },
]
`;

exports[`users suspended user 1`] = `
Array [
  Object {
    "model": "tms_User",
    "record": Object {
      "emailAddress": "customer@example.com",
      "inactive": true,
      "name": "The Customer",
      "source": "Zendesk",
      "uid": "111",
    },
  },
]
`;

exports[`users user 1`] = `
Array [
  Object {
    "model": "tms_User",
    "record": Object {
      "emailAddress": "customer@example.com",
      "inactive": false,
      "name": "The Customer",
      "source": "Zendesk",
      "uid": "111",
    },
  },
]
`;

exports[`zendesk process records from all streams 1`] = `
Array [
  "Processed 64 records",
  "Processed records by stream: {\\\\\\"mytestsource__zendesk__groups\\\\\\":2,\\\\\\"mytestsource__zendesk__satisfaction_ratings\\\\\\":2,\\\\\\"mytestsource__zendesk__tags\\\\\\":8,\\\\\\"mytestsource__zendesk__ticket_fields\\\\\\":9,\\\\\\"mytestsource__zendesk__ticket_metrics\\\\\\":14,\\\\\\"mytestsource__zendesk__tickets\\\\\\":14,\\\\\\"mytestsource__zendesk__users\\\\\\":15}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 120 records",
  "Would write records by model: {\\\\\\"faros_MetricDefinition\\\\\\":1,\\\\\\"faros_MetricValue\\\\\\":2,\\\\\\"faros_TmsTaskBoardOptions\\\\\\":1,\\\\\\"org_Team\\\\\\":2,\\\\\\"org_TeamMetric\\\\\\":2,\\\\\\"tms_Label\\\\\\":8,\\\\\\"tms_Project\\\\\\":1,\\\\\\"tms_Task\\\\\\":14,\\\\\\"tms_TaskAssignment\\\\\\":14,\\\\\\"tms_TaskBoard\\\\\\":1,\\\\\\"tms_TaskBoardProjectRelationship\\\\\\":1,\\\\\\"tms_TaskBoardRelationship\\\\\\":14,\\\\\\"tms_TaskDependency\\\\\\":1,\\\\\\"tms_TaskProjectRelationship\\\\\\":14,\\\\\\"tms_TaskTag\\\\\\":29,\\\\\\"tms_User\\\\\\":15}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
